### DISCUSSION TITLE
Late Initializers

### DISCUSSION THREAD

**Discussion by RandyBuchholz** [dotnet/csharplang#327] [General]
One of the things I miss when working with Tasks and DI is simple initialization.
```
var foo = new Bar() {
   a : "One",
   b : 2
}
```
It would be nice to be able to do this with injected things
```
var foo = _factory.GetNewBar() {
   a : "One",
   b : 2
}
```
or Task results
```
var foo = async GetNewBar() {
   a : "One",
   b : 2
}
```

**Comment by DavidArno** [comment]
I'm not opposing this idea, but isn't it just introducing a `with` statement to C#? It (probably) could just let me do:
```cs
var foo = new Bar();
foo {
   a : "One",
   b : 2
}
```
I'm sure I read recently that there are problems with `with` in VB.Net, but I may be imagining this.

**Comment by RandyBuchholz** [comment]
@DavidArno I think a difference is one line vs multiple lines, and expression body everything idea. 
```
public Bar Bar { get => _factory.GetNewBar() { a : "One", b : 2 }; }
```
or did you mean something like
```
var foo = new Bar() with
foo {
   a : "One",
   b : 2
};
```

**Comment by DavidArno** [comment]
No, I didn't mean that, so I guess my code isn't clear. With what you propose, I could do:
```cs
var foo = new Bar();
// some code
var bar = foo {
   a : "One",
   b : 2
}
```
But `foo` already have a reference to that instance of `Bar`, so I don't need a second one. I could there do:
```cs
var foo = new Bar();
// some code
_ = foo {
   a : "One",
   b : 2
}
// foo.a is now "One", foo.b is 2
```
But why have that `_ =`? Lose that and it becomes:
```cs
var foo = new Bar();
// some code
foo {
   a : "One",
   b : 2
}
```
At this point we have a `with`-like feature. So maybe just expand it to allow method invocations too:
```cs
var foo = new Bar();
// some code
foo {
   a : "One",
   b : 2,
   SomeMethod()
}
```
And now it's a complete `with` statement...
